Chore: Migrate to Freezed 3.0

davidpryor opened the pull request:
NOTE: This bumps the min SDK version to 3.0 but this seems to be planned already for the 9.0.0.

Description: Upgraded the freezed package to 3.0 and performed code migrations
Related Issue: #1295 



Summary:
- bumped package sdk min version to 3.0 for sealed classes and class modifiers features
- bumped example sdk min version to 3.0
- upgraded packages
- changed `WebPurchaseRedemptionResult` redirected constructors to be public so they may be used with `switch` expressions
- migrated from using `when` method to `switch` expressions
- fixed tests

Defuera commented:
Pls merge and release asap 🙏

g-apparence commented:
any news?

davidpryor commented:
> any news?
Just waiting on a review/comments from the maintainers

This probably wouldn't be able to ship until the next major release anyways as it bumps the minimum supported dart sdk.

tonidero commented on lib/models/web_purchase_redemption_result.dart:
```diff
@@ -7,14 +7,18 @@ import 'purchases_error.dart';
-  const factory WebPurchaseRedemptionResult.invalidToken() = _InvalidToken;
-  const factory WebPurchaseRedemptionResult.expired({required String obfuscatedEmail}) = _Expired;
+  const factory WebPurchaseRedemptionResult.success(
+      {required CustomerInfo customerInfo}) = Success;
```
Looks like there are some linter errors when running `flutter analyze lib` here. Do you mind fixing them?

davidpryor replied:
For sure! Took care of them 👍🏻 
```
➜ flutter analyze lib
Analyzing lib...                                                        
No issues found! (ran in 0.5s)
```

tonidero reviewed (commented):
Really sorry for the delay on this everyone! We are working on this now. The changes in this PR look great! However, as @davidpryor said, this is a breaking change, so we will release it as part of the next major 9.0.0. We will be releasing 9.0.0 stable soon, but in the meantime, we can release a new beta including these changes to unblock developers.

@davidpryor Would you mind updating these changes to the [9.0.0-dev](https://github.com/RevenueCat/purchases-flutter/tree/9.0.0-dev) branch?

davidpryor commented:
@tonidero Thanks for the comment!
I have gone ahead and rebased on 9.0.0-dev and pointed this PR to that branch.
I have also fixed the formatting errors that the analyze command found.

tonidero reviewed (approved):
Looking great thank you so much for doing this! We will be releasing a new beta of 9.0.0 with these changes. Stable release should arrive soon.
